.doom.d/config.org
#+TITLE: Org Mode Configuration
#+AUTHOR: Phillip O'Reggio
#+PROPERTY: header-args :emacs-lisp :tangle yes :comments link
#+STARTUP: hideblocks overview
#+BEGIN_SRC emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
#+END_SRC

* Preamble
Super early things
(Good for fixing/changing and bugfixes)
#+begin_src emacs-lisp
;; redefine this as eieio uses it but master depecrates it
(defun cl--generic-struct-tag (name &rest _)
  ;; Use exactly the same code as for `typeof'.
  `(if ,name (type-of ,name) 'null))
#+end_src

Functions used throughout the config

Only run the forms based on which OS

#+begin_src emacs-lisp
(defmacro when-macos (&rest forms)
  "Only run FORMS if OS is MacOS."
  (if IS-MAC
      `(mapc #'eval ',forms)))

(defmacro when-linux (&rest forms)
  "Only run FORMS if OS is Linux."
  (if IS-LINUX
      `(mapc #'eval ',forms)))
#+end_src
* General
** Private Information
Load some personal information
#+BEGIN_SRC emacs-lisp
(if (file-exists-p "~/.doom.d/Secret.el")
    (load! "Secret.el"))
#+END_SRC
** Library Functions
Functions used in other places in the configs
#+BEGIN_SRC emacs-lisp
;; Get a live xwidget buffer with url opened
(defun personal--xwidget-buffer (url)
  (xwidget-webkit-browse-url url)
  (let ((xwidget-buf (seq-find
                      (lambda (b) (eq (buffer-local-value 'major-mode b) 'xwidget-webkit-mode))
                      (buffer-list))))
    xwidget-buf))
#+END_SRC
** Font, Themes, and the General Look and Behavior
*** Fonts
Keeping all the font information in one place

Setting the main font
#+begin_src emacs-lisp
(when-macos
 (setq doom-font (font-spec :family "CozetteVector" :size 20)))

(when-linux
 (setq doom-font (font-spec :family "CozetteCrossedSevenVector" :size 18)))
;; (when-linux
;;  (setq doom-font (font-spec :family "More Perfect DOS VGA" :size 16)))
#+end_src
"CozetteCrossedSevenVector"
"More Perfect DOS VGA"

Make some faces use variable pitch
#+begin_src emacs-lisp
(custom-theme-set-faces
   'user
   '(line-number ((t (:inherit fixed-pitch))))
   '(line-numer-current-line ((t (:inherit fixed-pitch))))
   '(doom-modeline-buffer-major-mode ((t (:inherit fixed-pitch)))))
#+end_src

And change the font used for variable pitch fonts in org mode
#+begin_src emacs-lisp
(when-macos
 (setq personal-variable-pitch-font "Helvetica")
 (setq personal-fixed-pitch-font "MesloLGS Nerd Font"))

(when-linux
 (setq personal-variable-pitch-font "Roboto Mono")
;; (setq personal-fixed-pitch-font "CozetteCrossedSevenVector"))
 (setq personal-fixed-pitch-font "CozetteCrossedSevenVector"))

(set-face-attribute 'variable-pitch nil :family personal-variable-pitch-font)

(defun personal-fixup-fonts-org-mode ()
  (interactive)
  (set-face-attribute 'variable-pitch nil :family personal-variable-pitch-font)
  (set-face-attribute 'org-modern-symbol nil :family personal-fixed-pitch-font)
  (set-face-attribute 'org-block nil :family personal-fixed-pitch-font)
  (set-face-attribute 'org-table nil :family personal-fixed-pitch-font))

(after! org-modern
  (setq org-modern-table nil)
  (personal-fixup-fonts-org-mode))
#+end_src


Make fixed pitch not be a constant size, for Demap and such so certain fonts (comments in
particular) don't appear super large compared to everything else
#+begin_src emacs-lisp
(defun personal-make-fixed-pitch-shrink ()
  (interactive)
  (set-face-attribute 'fixed-pitch nil :height 'unspecified))

(personal-make-fixed-pitch-shrink)
(add-hook! prog-mode #'personal-make-fixed-pitch-shrink)
(add-hook! doom-load-theme-hook #'personal-make-fixed-pitch-shrink)
#+end_src

*** Look
Set ewal color location json
#+begin_src emacs-lisp
(after! ewal
  (setq ewal-json-file "~/.cache/wallust/colors.json"))
#+end_src

Set the color theme and font
#+BEGIN_SRC emacs-lisp
(when-macos
(setq doom-theme 'starry-transparent))

(when-linux
(setq doom-theme 'ewal-doom-vibrant))

;; Make comments brighter if using ewal
(setq ewal-doom-vibrant-brighter-comments t)
(setq ewal-doom-vibrant-comment-bg nil)

(when-linux
 (defun personal-make-ewal-comment-brighter ()
   (interactive)
   (set-face-attribute 'font-lock-comment-face nil :foreground (ewal-load-color 'green -0))
   (set-face-attribute 'org-block-begin-line nil :foreground (ewal-load-color 'green -0))
   (set-face-attribute 'org-block-end-line nil :foreground (ewal-load-color 'green -0))))
#+END_SRC

Cursor color
#+begin_src emacs-lisp
(setq evil-normal-state-cursor '(box "#528AF2")
      evil-insert-state-cursor '(bar "#528AF2")
      evil-visual-state-cursor '(hollow "#FDDC87")
      evil-operator-state-cursor '(evil-half-cursor "#528AF2")
      evil-replace-state-cursor '(hbar "#FDDC87"))
#+end_src

Make the emacs frame transparent because I like my laptop background 🙃
#+BEGIN_SRC emacs-lisp
(defun personal-set-alpha-background-frame-params ()
  (let ((bg-transparency 0.75))
    (set-frame-parameter (selected-frame) 'alpha-background bg-transparency)
    (add-to-list 'default-frame-alist `(alpha-background . ,bg-transparency))))

(add-hook 'doom-load-theme-hook #'personal-set-alpha-background-frame-params)
#+END_SRC

And some functions to manipulate transparency on the frame
#+begin_src emacs-lisp
(defun personal-set-alpha-background-on-frame ()
  (interactive)
  (let ((alpha (read-number "Enter transparency (0.0-1.0): ")))
    (set-frame-parameter (selected-frame) 'alpha-background alpha)))
#+end_src

Control size of emacs on startup (mainly for macos since I use WM on linux)
#+begin_src emacs-lisp
(when-macos
 (setq initial-frame-alist '((top . 100) (left . 400) (width . 100) (height . 40))))
#+END_SRC

Set cursor color
#+begin_src emacs-lisp
;; TODO
;; (evil-set-cursor-color (ewal-get-color cyan))
#+end_src

Show Line numbers
#+BEGIN_SRC emacs-lisp
(setq display-line-numbers-type t)
#+END_SRC

Modify the appearance of the mode-line
#+BEGIN_SRC elisp
(setq doom-modeline-major-mode-icon t)
#+END_SRC

*** Behavior
Add a fill column to prevent long lines (116 is the width of the vertical monitor)
#+BEGIN_SRC emacs-lisp
(setq-default fill-column 100)
#+END_SRC

and automatically break long lines in source code buffers and org mode files
#+BEGIN_SRC emacs-lisp
(add-hook! prog-mode (auto-fill-mode t))
#+END_SRC

Show a message when doing garbage collection
#+BEGIN_SRC emacs-lisp
;; (setq garbage-collection-messages t)
#+END_SRC

Have evil-mode treat emacs /symbols/ as words instead of by character
(symbol definition changes based on the language)
#+BEGIN_SRC emacs-lisp
(with-eval-after-load 'evil
    (defalias #'forward-evil-word #'forward-evil-symbol)
    ;; make evil-search-word look for symbol rather than word boundaries
    (setq-default evil-symbol-word-search t))

#+END_SRC

Enable pixel scrolling precision
#+BEGIN_SRC emacs-lisp
(pixel-scroll-precision-mode 1)
(setq pixel-scroll-precision-use-momentum t)
#+END_SRC

Let search look at invisible content (useful in org mode)
#+begin_src emacs-lisp
(setq search-invisible t)
#+end_src

** Splash Screen Edits
Function for making random wave art for the banner
#+begin_src emacs-lisp :lexical t

(setq personal-wave-param-low 1)
(setq personal-wave-param-high 100)

(defun personal-frac-function (x)
  (if (>= x 0)
      (- x (floor x))
    (* -1 (- (* -1 x) (floor (* -1 x))))))

(defun personal-wave-function-1 (x y a b c d)
  (let* ((highest-magnitude (+ 1 1))
         (wave-result (+ (sin (* (/ a b) x)) (sin (* (/ c d 1.0) y))))
         (normalized (/ (+ wave-result highest-magnitude) (* 2 highest-magnitude))))
    normalized))

(defun personal-wave-function-2 (x y a b c d)
  (let* ((highest-magnitude 1)
         (wave-result (sin (* x y (/ a c 1.0) (/ b d 1.0))))
         (normalized (/ (+ wave-result highest-magnitude) (* 2 highest-magnitude))))
    normalized))

(defun personal-wave-function-3 (x y a b c d)
  (let* ((highest-magnitude 1)
         (wave-result (sin (* x y (* (/ a 1) c 1.0) (/ b d 1.0))))
         (normalized (/ (+ wave-result highest-magnitude) (* 2 highest-magnitude))))
    normalized))

(defun personal-wave-function-4 (x y a b c d)
  (let* ((highest-magnitude 1)
         (wave-result (cos (* (log (* (- (* 2 pi) x) c )) (log (* (- (* 2 pi) y) b)) a d 0.01)))
         (normalized (/ (+ wave-result highest-magnitude) (* 2 highest-magnitude))))
    normalized))

(defun personal-wave-function-5 (x y a b c d)
  (let* ((highest-magnitude 1)
         (base-number (personal-frac-function
                       (+
                        (* (* 0.8 b) (sin (* a y)))
                        (* (* 0.8 d) (sin (* c x))))))
         (wave-result (* 2 (- base-number 0.5)))
         (normalized (/ (+ wave-result highest-magnitude) (* 2 highest-magnitude))))
    normalized))

(personal-wave-function-3 1 1 1 1 1 1)

(defun personal-choose-wave-func (wave-funcs)
  "Choose random function from WAVE-FUNCS"
  (nth (random (length wave-funcs)) wave-funcs))

(defun personal-intensity-to-character (intensity)
  "Maps INTENSITY to a character from .,-~:;=!*#$@"
  (pcase intensity
    ((pred (lambda (x) (<= x (* 1 0.077))))
     ?\s)
    ((pred (lambda (x) (<= x (* 2 0.077))))
     ?\s)
    ((pred (lambda (x) (<= x (* 3 0.077))))
     ?\s)
    ((pred (lambda (x) (<= x (* 4 0.077))))
     ?.)
    ((pred (lambda (x) (<= x (* 5 0.077))))
     ?-)
    ((pred (lambda (x) (<= x (* 6 0.077))))
     ?:)
    ((pred (lambda (x) (<= x (* 7 0.077))))
     ?\;)
    ((pred (lambda (x) (<= x (* 8 0.077))))
     ?=)
    ((pred (lambda (x) (<= x (* 9 0.077))))
     ?!)
    ((pred (lambda (x) (<= x (* 10 0.077))))
     ?*)
    ((pred (lambda (x) (<= x (* 11 0.077))))
     ?#)
    ((pred (lambda (x) (<= x (* 12 0.077))))
     ?$)
    ((pred (lambda (x) (<= x (* 13 0.077))))
     ?@)
    (x
     ?\@)))

(defun personal-scale-to-2pi (cur max)
  "Scales a number CUR between [0, max) to [0, 2pi)"
  (* (/ cur max 1.0) (* float-pi 2)))

(defun personal-make-ascii-wave-art (height width)
  (let ((output-lines (mapcar (lambda (c) (make-string width c)) (make-list height ?.)))
        (wave-func (personal-choose-wave-func
                    '(personal-wave-function-1 personal-wave-function-2 personal-wave-function-3
                      personal-wave-function-4 personal-wave-function-5)))
        (a (+ (random (- personal-wave-param-high personal-wave-param-low)) personal-wave-param-low))
        (b (+ (random (- personal-wave-param-high personal-wave-param-low)) personal-wave-param-low))
        (c (+ (random (- personal-wave-param-high personal-wave-param-low)) personal-wave-param-low))
        (d (+ (random (- personal-wave-param-high personal-wave-param-low)) personal-wave-param-low)))
    (cl-loop for i from 0 to (- height 1) do
             (cl-loop for j from 0 to (- width 1) do
                      (let* ((x (personal-scale-to-2pi j width))
                             (y (personal-scale-to-2pi i height))
                             (intensity (apply wave-func `(,x ,y ,a ,b ,c ,d)))
                             (character (personal-intensity-to-character intensity)))
                        (aset (nth i output-lines) j character))))

    output-lines))

(after! +doom-dashboard-mode-hook
  (personal-make-ewal-comment-brighter))
#+end_src

Custom Banner text
#+begin_src emacs-lisp :lexical t
(setq personal-wave-width 60)
(setq personal-wave-height 30)

(defun personal-custom-splash-ascii ()
  (let* ((banner
          (personal-make-ascii-wave-art personal-wave-height personal-wave-width))
         (longest-line (apply #'max (mapcar #'length banner))))
    (put-text-property
     (point)
     (dolist (line banner (point))
       (insert (+doom-dashboard--center
                +doom-dashboard--width
                (concat line (make-string (max 0 (- longest-line (length line))) 32)))
               "\n"))
     'face 'doom-dashboard-banner)))

(setq +doom-dashboard-ascii-banner-fn #'personal-custom-splash-ascii)
#+end_src

Remove splash screen menu options I don't use
#+begin_src emacs-lisp

(defun personal-change-doom-dashboard-options ()
  "Change dashboard options for start screen"
  (require 'nerd-icons)
  (after! nerd-icons
    (setq +doom-dashboard-menu-sections
          '(("Resume" :icon
             (nerd-icons-octicon "nf-oct-history" :face 'doom-dashboard-menu-title)
             :when
             (cond
              ((featurep! :ui workspaces)
               (file-exists-p
                (expand-file-name persp-auto-save-fname persp-save-dir)))
              ((require 'desktop nil t)
               (file-exists-p
                (desktop-full-file-name))))
             :face
             (:inherit
              (doom-dashboard-menu-title bold))
             :action doom/quickload-session)
            ("Config" :icon
             (nerd-icons-octicon "nf-oct-tools" :face 'doom-dashboard-menu-title)
             :when
             (file-directory-p doom-private-dir)
             :action doom/open-private-config)))))

(personal-change-doom-dashboard-options)
#+end_src

Last Crash Info (for MacOS)
(unused now since I don't use Emacs on MacOS much
#+begin_src emacs-lisp
;; (defun personal--last-crash-delta (crash-log-dir emacs-crash-cache-file)
;;   "Computes time since last crash.
;; Return result as a numeric number that needs to be converted to human readable string using `ts-human-format-duration'.
;; 
;; CRASH-LOG-DIR is the directory where Emacs crash log is located (used to run touch).
;; 
;; EMACS-CRASH-CACHE-file is a file (may not exist yet) that stores the time of the latest crash. Is
;; used when the CRASH-LOG-DIR has no information for the last crash time, and is created/updated on
;; each parse."
;;   (require 'ts)
;;   (let* ((dir-name (expand-file-name crash-log-dir))
;;          (last-crash-log (shell-command-to-string (format "ls -r %s | rg -i emacs | head -1"
;;                                                           dir-name))))
;;     (if (length= last-crash-log 0)
;;         (personal--use-crash-cache (expand-file-name emacs-crash-cache-file))
;;       (personal--use-crash-log
;;        (expand-file-name crash-log-dir)
;;        (expand-file-name emacs-crash-cache-file)))))
;; 
;; (defun personal--use-crash-cache (emacs-crash-cache-file)
;;   (unless (file-exists-p emacs-crash-cache-file)
;;     (write-region (shell-command-to-string "date -R") nil emacs-crash-cache-file))
;;   (let ((last-crash-string (with-temp-buffer
;;                         (insert-file-contents emacs-crash-cache-file)
;;                         (buffer-string))))
;;     (personal--compute-delta last-crash-string)))
;; 
;; (defun personal--use-crash-log (log-file cache-file)
;;   (let ((last-crash-string
;;          (shell-command-to-string (format "date -r %s -R" log-file))))
;;     (write-region last-crash-string nil cache-file)
;;     (personal--compute-delta last-crash-string)))
;; 
;; (defun personal--compute-delta (last-crash-time-string)
;;   (ts-diff (ts-parse (shell-command-to-string "date -R")) (ts-parse last-crash-time-string)))
;; 
;; (defun personal--is-longest-no-crash-time (delta best-time-file)
;;   "Return `t' if DELTA is bigger than the number in BEST-TIME-FILE.
;; Also handles updating the number if it is bigger, or creating it if it doesn't exist."
;;   (let ((best-time-file-name (expand-file-name best-time-file)))
;;     (unless (file-exists-p best-time-file-name)
;;       (write-region "0" nil best-time-file-name))
;;     (let ((best-delta (string-to-number (with-temp-buffer
;;                         (insert-file-contents best-time-file-name)
;;                         (buffer-string)))))
;;       (if (> delta best-delta)
;;           (progn
;;             ;; write that to file instead
;;             (write-region (number-to-string delta) nil best-time-file-name)
;;             ;; return t
;;             t)
;;         nil))))
;; 
;; 
;; (defun doom-last-crash-line ()
;;   "Say how long since Emacs last crashed.
;; BEST-TIME-NAME is optional and specifies absolute path to file that contains the longest time Emacs"
;;   (let* ((delta (personal--last-crash-delta "~/Library/Logs/DiagnosticReports" "~/.doom.d/splash-last-crash.txt"))
;;          (delta-string (ts-human-format-duration delta)))
;;     (if (personal--is-longest-no-crash-time delta "~/.doom.d/splash-longest-last-crash.txt")
;;         (format "%s since last incident! (NEW RECORD!)" delta-string)
;;       (format "%s since last incident!" delta-string))))
;; 
;; (defun doom-dashboard-phrase ()
;;   "Get a splash phrase, flow it over multiple lines as needed, and make fontify it."
;;   (mapconcat
;;    (lambda (line)
;;      (+doom-dashboard--center
;;       +doom-dashboard--width
;;       (with-temp-buffer
;;         (insert-text-button
;;          line
;;          'action
;;          (lambda (_) (+doom-dashboard-reload t))
;;          'face 'doom-dashboard-menu-title
;;          'mouse-face 'doom-dashboard-menu-title
;;          'help-echo "Last crash"
;;          'follow-link t)
;;         (buffer-string))))
;;    (split-string
;;     (with-temp-buffer
;;       (insert (doom-last-crash-line))
;;       (setq fill-column (min 70 (/ (* 2 (window-width)) 3)))
;;       (fill-region (point-min) (point-max))
;;       (buffer-string))
;;     "\n")
;;    "\n"))
;; 
;; ;; Only place this message if ts is available (avoid startup errors)
;; (when (require 'ts nil 'noerror)
;;   (defadvice! doom-dashboard-widget-loaded-with-phrase ()
;;     :override #'doom-dashboard-widget-loaded
;;     (setq line-spacing 0.2)
;;     (insert
;;      "\n\n"
;;      (propertize
;;       (+doom-dashboard--center
;;        +doom-dashboard--width
;;        (doom-display-benchmark-h 'return))
;;       'face 'doom-dashboard-loaded)
;;      "\n"
;;      (doom-dashboard-phrase)
;;      "\n")))
;; 
;; (if IS-MAC
;;     (add-hook 'doom-load-theme-hook #'personal-change-doom-dashboard-options))
#+end_src

** Useful Functions
*** New Functions
Toggle the transparency of emacs
#+BEGIN_SRC emacs-lisp
;; Initialize transparency to `true`

;; TODO change this to not use global state, but instead use frame-local state

(put 'cfg-transparency 'state t)

(defun personal-toggle-transparency ()
  "Toggle the transparency of emacs"
  (interactive)
  (if (get 'cfg-transparency 'state)
      (progn
        (set-frame-parameter (selected-frame) 'alpha '(100 100))
        (put 'cfg-transparency 'state nil))
    (progn
      (set-frame-parameter (selected-frame) 'alpha '(85 85))
      (put 'cfg-transparency 'state t))
    ))
#+END_SRC

Control size of Emacs frame:
#+BEGIN_SRC emacs-lisp
(defun personal-frame-change-size (width height)
  "Modify size of window frame by increasing it by WIDTH and HEIGHT."
  (let ((cur-width (frame-width (window-frame)))
        (cur-height (frame-height (window-frame))))
    (set-frame-size (window-frame) (+ cur-width width) (+ cur-height height))))

(defun personal-frame-full-screen ()
  "Toggle frame to be fullscreen."
  (interactive)
  (toggle-frame-fullscreen))

(defun personal-frame-increase-width ()
  "Increase frame width by 1."
  (interactive)
  (personal-frame-change-size 1 0))

(defun personal-frame-decrease-width ()
    "Decrease frame width  by 1."
    (interactive)
    (personal-frame-change-size -1 0))

(defun personal-frame-increase-height ()
    "Increase frame height by 1."
    (interactive)
    (personal-frame-change-size 0 1))

(defun personal-frame-decrease-height ()
    "Decrease frame height by 1."
    (interactive)
    (personal-frame-change-size 0 -1))

(defun personal-frame-increase-diag ()
    "Increase frame width and height by 1."
    (interactive)
    (personal-frame-change-size 1 1))

(defun personal-frame-decrease-diag ()
    "Decrease frame width and height by 1."
    (interactive)
    (personal-frame-change-size -1 -1))
#+END_SRC

Control position of emacs frame:
#+BEGIN_SRC emacs-lisp
(defun personal-frame-move (down right)
  "Move window frame by DOWN and RIGHT."
  (pcase (frame-position)
    (`(,x . ,y) (set-frame-position (selected-frame) (+ x right) (+ y down)))))

(defun personal-move-frame-down-30 ()
  "Move window frame down 30."
  (interactive)
  (personal-frame-move 30 0))

(defun personal-move-frame-down-5 ()
  "Move window frame down 5."
  (interactive)
  (personal-frame-move 5 0))

(defun personal-move-frame-up-30 ()
  "Move window frame up 30."
  (interactive)
  (personal-frame-move -30 0))

(defun personal-move-frame-up-5 ()
  "Move window frame down 5."
  (interactive)
  (personal-frame-move -5 0))

(defun personal-move-frame-left-30 ()
  "Move window frame left 30."
  (interactive)
  (personal-frame-move 0 -30))

(defun personal-move-frame-left-5 ()
  "Move window frame left 5."
  (interactive)
  (personal-frame-move 0 -5))

(defun personal-move-frame-right-30 ()
  "Move window frame right 30."
  (interactive)
  (personal-frame-move 0 30))

(defun personal-move-frame-right-5 ()
  "Move window frame right 5."
  (interactive)
  (personal-frame-move 0 5))
#+END_SRC

Change fill-column
#+BEGIN_SRC elisp
#+END_SRC

Open up terminal in narrow vertical split
#+BEGIN_SRC emacs-lisp
(defun personal-sterm ()
  "Opens a terminal in a split on the left"
  (interactive)
  (require 'vterm)
  (progn
    (split-window-right 45)
    (+vterm/here t)))
#+END_SRC

Open up google in a split
#+BEGIN_SRC elisp
(defun personal-google-split ()
  "Open google in vertical split using xwidget-webkit"
  (interactive)
  (let ((google-url "https://www.google.com")
        (xwidget-buffer (lambda ()
                          (seq-find
                           (lambda (b) (eq (buffer-local-value 'major-mode b) 'xwidget-webkit-mode))
                           (buffer-list)))))
    (split-window-right)
    (xwidget-webkit-browse-url google-url)
    (switch-to-buffer (funcall xwidget-buffer))))

(defun personal-google-here ()
  "Open google in current buffer"
  (interactive)
  (let ((google-url "https://www.google.com"))
    (personal--xwidget-buffer google-url)))
#+END_SRC

Open google in window with xwidget
#+BEGIN_SRC elisp
(defun personal-open-google ()
  "Open google in in window using xwidget-webkit"
  (interactive)
  (let ((google-url "https://www.google.com"))
    (xwidget-webkit-browse-url google-url)))
#+END_SRC

*** Changing Old Ones
nothing yet

** Keybindings and Custom Commands
*** Keybindings
jk -> Escape
#+begin_src emacs-lisp
(require 'corfu)
(require 'key-chord)

(defun personal-jk-escape ()
  (interactive)
  (corfu-quit)
  (evil-normal-state))

(key-chord-mode 1)
(key-chord-define evil-insert-state-map  "jk" #'personal-jk-escape)
#+end_src

Using Doom Evil
#+begin_src emacs-lisp
(setq-default evil-escape-key-sequence "jk")
#+end_src

Make moving around splits as easy as pressing space.
#+BEGIN_SRC emacs-lisp
(map! :leader :mode 'global
  "h" #'evil-window-left
  "l" #'evil-window-right
  "j" #'evil-window-down
  "k" #'evil-window-up

  "H" #'+evil/window-move-left
  "L" #'+evil/window-move-right
  "K" #'+evil/window-move-up
  "J" #'+evil/window-move-down
  )
#+END_SRC

Use dirvish over find-file
#+BEGIN_SRC emacs-lisp
(defun personal-dirvish-find-file (filename &optional wildcards)
  (interactive
   (find-file-read-args "🔍 Open: "
                        (confirm-nonexistent-file-or-buffer)))
  (dirvish-dwim filename))

(map! :leader :mode 'global
 "." #'personal-dirvish-find-file
  )
#+END_SRC

Change =;= to =:= (for vim ex mode)
#+BEGIN_SRC emacs-lisp
(map! :nv
  ";" 'evil-ex
  )
#+END_SRC

Map "s" to the sneak motion
#+BEGIN_SRC emacs-lisp
;; Unbind "s" from everything else
(map!
 :map evil-normal-state-map
 "s" nil
 "S" nil)

;; Then bind it
(map!
 :nv "s" #'evil-avy-goto-char-2-below
 :n "S" #'evil-avy-goto-char-2-above)
#+END_SRC

*** Custom ex Commands
~:G~ for magit status
#+BEGIN_SRC emacs-lisp
(evil-ex-define-cmd "G" #'magit-status)
#+END_SRC

~:S~ to search google and ~:SS~ to open google in current frame
#+BEGIN_SRC emacs-lisp
(evil-ex-define-cmd "S" #'personal-google-split)
(evil-ex-define-cmd "SS" #'personal-google-here)
#+END_SRC

~:E~ to search google with eww
#+BEGIN_SRC emacs-lisp
(defun personal-eww-google ()
  (interactive)
  (let* ((query (read-from-minibuffer "Search for: "))
         (url-extension (replace-regexp-in-string " " "+" query )))
  (eww (concat "https://www.google.com/search?q=" url-extension))))

(evil-ex-define-cmd "E" #'personal-eww-google)
#+END_SRC

~:Fexplore~ to open a project drawer
#+BEGIN_SRC emacs-lisp
(evil-ex-define-cmd "Fexplore" #'treemacs)
#+END_SRC

~:PRReview~ to diff between two branches
#+BEGIN_SRC emacs-lisp
(evil-ex-define-cmd "PRReview" #'magit-diff-range)
#+END_SRC

~:STerm~ to open terminal in a split
#+BEGIN_SRC emacs-lisp
(evil-ex-define-cmd "STerm" #'personal-sterm)
#+END_SRC

** Memory Usage
Let Emacs use more memory for reading and garbage collection
#+begin_src elisp
;; 100 mb
(setq gc-cons-threshold 100000000)
(setq read-process-output-max (* 1024 1024)) ;; 1mb
#+end_src

** Fixing Things
A place for hacks and functions to fix fires

(nothing right now)

** Abbrevs
Manage abbreviations
#+BEGIN_SRC emacs-lisp
(quietly-read-abbrev-file (expand-file-name "~/.doom.d/abbrev.el"))
#+END_SRC

* Modes
** Text Mode
Mode for human readable text

Recommend words when typing in text mode files:
#+BEGIN_SRC emacs-lisp
(defun personal-setup-completion-for-text-mode ()
  (interactive)
  (setq-local
   completion-at-point-functions (list (cape-capf-super
                                        #'cape-dict
                                        #'cape-dabbrev
                                        #'cape-emoji) #'cape-file)))

(add-hook! text-mode #'personal-setup-completion-for-text-mode)
#+END_SRC

** Org Mode
*** Look
Use org modern mode
#+BEGIN_SRC emacs-lisp
(add-hook 'org-agenda-finalize-hook #'org-modern-agenda)
#+END_SRC

Set the org directory for org related files
#+BEGIN_SRC emacs-lisp
(setq org-directory "~/Dropbox/agenda")
#+END_SRC

Make org-mode documents look a little nicer by hiding markers and showing symbols
#+BEGIN_SRC emacs-lisp
(setq org-hide-emphasis-markers t)
(setq org-pretty-entities t)
#+END_SRC

#+BEGIN_SRC emacs-lisp
;; (setq
;;     org-superstar-headline-bullets-list '("⁖" "◉" "○" "✸" "✿")
;; )
#+END_SRC

_Snippets to prettify Org mode based on this:_ [[https://zzamboni.org/post/beautifying-org-mode-in-emacs/][Beautifying Org Mode in Emacs]]:

Make org-mode files use variable pitch fonts to look more like text documents
([[*Fonts][See this]])

Files can opt out by having this at the *end* of the file:
#+BEGIN_EXAMPLE org
;; Local Variables:
;; eval: (variable-pitch-mode 0)
;; End:
#+END_EXAMPLE

*** Functionality
Add =proselint= to lint org-mode
#+begin_src elisp
(setq flycheck-proselint-executable "~/Library/Python/3.8/bin/proselint")
#+end_src

Setup =org-download=, which makes the process of putting images into orgmode much easier
Put downloaded images into an =images= directory and include timestamp. Use the command =pngpaste= to get the image
from the clipboard.
#+begin_src emacs-lisp
(after! org-download
  (setq org-download-method 'directory)
  (setq-default org-download-image-dir "images")
  (setq org-download-timestamp "%Y%m%d-%H%M%S_")
  (setq org-download-screenshot-method "/usr/local/bin/pngpaste %s"))
#+end_src

Advice =org-fancy-priorities-mode= which sometimes errors since =org-download= hasn't loaded yet
#+begin_src emacs-lisp
(defadvice! personal--load-org-download-with-fancy-priorities ()
  :before #'org-fancy-priorities-mode
  (require 'org-download))
#+end_src

Allow pasting of images with control over the width, and bind it to keybind
#+begin_src emacs-lisp
(defun org-download-screenshot-with-size ()

.doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(load! "Secret.el")

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
(setq doom-font (font-spec :family "MesloLGS Nerd Font" :size 16))

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-city-lights)

;; If you use `org' and don't want your org files in the default location below,;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)


;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys ;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c g k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c g d') to jump to their definition and see how
;; they are implemented.

;; --- Default window size on startup
(setq initial-frame-alist '((top . 50) (left . 240) (width . 120) (height . 45)))

;; --- Keymaps
(map! :leader :mode 'global
  "h" #'evil-window-left
  "l" #'evil-window-right
  "j" #'evil-window-down
  "k" #'evil-window-up
  "gt" '+workspace/cycle
  )
(map! :nv
  ";" 'evil-ex
  )
(map! :leader :g
      "gt" #'+workspace/cycle)

(require 'evil-snipe)
(setq evil-snipe-scope 'buffer)
(evil-snipe-mode +1)

;; --- Company config
(setq company-idle-delay 0.1
      company-minimum-prefix-length 3)

(use-package lsp-sourcekit
  :after lsp-mode
  :config
  (setq lsp-sourcekit-executable "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/sourcekit-lsp"))
(use-package swift-mode
  :hook (swift-mode . (lambda () (lsp))))

(use-package company-box
  :hook (company-mode . company-box-mode))

;; Company Completion
(set-company-backend! '(text-mode org-mode) 'company-ispell 'company-dabbrev)
(set-company-backend! 'swift-mode 'company-capf 'company-dabbrev)

;; --- AMPL editing
(setq auto-mode-alist
      (cons '("\\.mod$" . ampl-mode) auto-mode-alist))
(setq auto-mode-alist
      (cons '("\\.dat$" . ampl-mode) auto-mode-alist))
(setq auto-mode-alist
      (cons '("\\.ampl$" . ampl-mode) auto-mode-alist))
(setq interpreter-mode-alist
      (cons '("ampl" . ampl-mode)
            interpreter-mode-alist))
(autoload 'ampl-mode "ampl-mode" "Ampl editing mode." t)

;; --- Latex and PDFs
(latex-preview-pane-enable)

;; Retina display / less fuzzy pdfs
(setq pdf-view-use-scaling t pdf-view-use-imagemagick nil)

;; --- Org mode
(setq org-hide-emphasis-markers t)
(setq org-pretty-entities t)
;; Fix bug(?) where newline is set to void in orgmode
(add-hook 'org-mode-hook (lambda () (electric-indent-local-mode -1)))

;; --- Fill column
(setq-default fill-column 116)

;; --- Set transparency
(set-frame-parameter (selected-frame) 'alpha '(85 85))
(add-to-list 'default-frame-alist '(alpha 85 85))
